crypto/rand/rand_lib.c

```diff
@@ -211,25 +220,308 @@ void rand_cleanup_int(void)
+{
+    return pool->entropy;
+}
+
+/*
+ * Return the |pool|'s buffer length to the caller.
+ */
+size_t RAND_POOL_length(RAND_POOL *pool)
+{
+    return pool->len;
+}
+
+/*
+ * Detach the |pool| buffer and return it to the caller.
+ * It's the responsibility of the caller to free the buffer
+ * using OPENSSL_secure_free().
+ */
+unsigned char *RAND_POOL_detach(RAND_POOL *pool)
+{
+    unsigned char *ret = pool->buffer;
+    pool->buffer = NULL;
+    return ret;
+}
+
+
+/*
+ * If every byte of the input contains |entropy_per_bytes| bits of entropy,
+ * how many bytes does one need to obtain at least |bits| bits of entropy?
+ */
+#define ENTROPY_TO_BYTES(bits, entropy_per_bytes) \
+    (((bits) + ((entropy_per_bytes) - 1))/(entropy_per_bytes))
+
+
+/*
+ * Checks whether the |pool|'s entropy is available to the caller.
+ * This is the case when entropy count and buffer length are high enough.
+ * Returns
+ *
+ *  |entropy|  if the entropy count and buffer size is large enough
+ *      0      otherwise
+ */
+size_t RAND_POOL_entropy_available(RAND_POOL *pool)
 {
-    return RAND_poll_ex(call_rand_add, NULL);
+    if (pool->entropy < pool->requested_entropy)
+        return 0;
+
+    if (pool->len < pool->min_len)
+        return 0;
+
+    return pool->entropy;
+}
+
+/*
+ * Returns the (remaining) amount of entropy needed to fill
+ * the random pool.
+ */
+
+size_t RAND_POOL_entropy_needed(RAND_POOL *pool)
+{
+    if (pool->entropy < pool->requested_entropy)
+        return pool->requested_entropy - pool->entropy;
+
+    return 0;
+}
+
+/*
+ * Returns the number of bytes needed to fill the pool, assuming
+ * the input has 'entropy_per_byte' entropy bits per byte.
+ * In case of an error, 0 is returned.
+ */
+
+size_t RAND_POOL_bytes_needed(RAND_POOL *pool, unsigned int entropy_per_byte)
+{
+    size_t bytes_needed;
+    size_t entropy_needed = RAND_POOL_entropy_needed(pool);
+
+    if (entropy_per_byte < 1 || entropy_per_byte > 8) {
+        RANDerr(RAND_F_RAND_POOL_BYTES_NEEDED, RAND_R_ARGUMENT_OUT_OF_RANGE);
+        return 0;
+    }
+
+    bytes_needed = ENTROPY_TO_BYTES(entropy_needed, entropy_per_byte);
+
+    if (bytes_needed > pool->max_len - pool->len) {
+        /* not enough space left */
+        RANDerr(RAND_F_RAND_POOL_BYTES_NEEDED, RAND_R_RANDOM_POOL_OVERFLOW);
+        return 0;
+    }
+
+    if (pool->len < pool->min_len &&
+        bytes_needed < pool->min_len - pool->len)
+        /* to meet the min_len requirement */
+        bytes_needed = pool->min_len - pool->len;
+
+    return bytes_needed;
+}
+
+/* Returns the remaining number of bytes available */
+size_t RAND_POOL_bytes_remaining(RAND_POOL *pool)
+{
+    return pool->max_len - pool->len;
+}
+
+/*
+ * Add random bytes to the random pool.
+ *
+ * It is expected that the |buffer| contains |len| bytes of
+ * random input which contains at least |entropy| bits of
+ * randomness.
+ *
+ * Return available amount of entropy after this operation.
```

Comment:
This return value only seems to be used once, to just return it. Since it just calls an other function to return this, wouldn't it be better to just let that one user just call RAND_POOL_entropy_available() itself?

Reply:
I'm a bit torn on this one.  I can see the usefulness of returning the currently available entropy and would suggest doing this rather than returning the tri-state indication of _full_, _not full_ and _error_.

However, I'm having difficulty thinking of a case where doing this is beneficial -- the caller is quite able to call `RAND_POOL_entropy_available()` itself if it needs to quantify the availability (which won't be often).  The rest of the time a success or fail result should be enough.

I think it should either be left as it is (return the available entropy) or simplified (return a success/failure boolean).  I lean towards the former but not strongly.

Reply:
Most applications will not be able to use the return value correctly.  I think you consider not providing one and just provide an int and an ERR_ value pushed on error.

Reply:
The RAND_POOL API is not intended to be used by 'most applications'. It was designed for the 'backend' where random input is pulled from entropy sources. So this argument does not count. Certainly, if the design of the RAND_POOL API needs some improvements in order to fulfil its purpose, I am open for suggestions. But for this pull request the discussion comes too late. I will only change nits and errors, but don't start any design changes anymore. Otherwise this pull requests will become a never ending story.

If I would change the return value of RAND_POOL_add() now, other functions like RAND_POOL_acquire_entropy() and rand_acquire_entropy_from_xxx() would have to be changed accordingliy, because currently they return the available entropy count, too, very often simply by forwarding the return value of RAND_POOL_add().